cEP-0023.md

```diff
@@ -0,0 +1,277 @@
+# Breaking down coala
+
+| Metadata |                                               |
+|----------|-----------------------------------------------|
+| cEP      | 0023                                          |
+| Version  | 1.0                                           |
+| Title    | Breaking down the coala                       |
+| Authors  | Muhammad Kaisar Arkhan <mailto:yuki@coala.io> |
+| Status   | Proposed                                      |
+| Type     | Feature                                       |
+
+## Abstract
+
+This cEP proposes a new architecture evolution for coala.
+
+## How coala works
+
+coala is a Python program that also provides API for bears and interacting with
+them.
+
+Bears are Python classes filled with metadata needed to execute linter programs
+and may contain helper code to generate configuration files. Since bears are
+Python code, some bears have the actual linter inside of them and some just
+execute the Python API of a linter.
+
+Bears are written with the API provided by the coala core and they're running
+under the same process as coala.
+
+On startup, coala reads and intepret the configuration file. Configuration files
+are separated by sections which contains what bear is required and the
+configuration for the bear.
+
+After it's intepreted, it is passed over to an executor which will executes it
+per section. The executor will find and import the bear class which usually
+resides in the Python packages path. It then pass the Section data to the bear
+and set it's require configuration. The bear first prepare the whole command
+along with it's parameters or generate a configuration file. After that's done,
+the linter program will be executed by the bear as a separate process and the
+bear will intepret it's standard output by matching a regex. The bear will yield
+the output of the program to the executor which will end up to the generator
+where the user see the warning/error and determine an action.
+
+## The Architectural Issue
+
+There is no clear distinction between the coala core which job is to intepret
+and run bears and the one that shows the command line interface to the user. In
+addition, bears in essence can do anything it wants to since it is just a Python
+class that is loaded into the coala process.
+
+This caused numerous unwanted regressions between bears and coala and costed a
+lot in maintenance. This is also the reason why bears are not released
+separately as a "collection" rather it is shipped with coala core along with
+every linter in the collection.
+
+Instead of leaving the project at this state, we should define borders and setup
+a more proper architecture and standards.
+
+## The new architecture
+
+Bears are now just simple formatted plain-text file containing the metadata
+needed. It can also contain a simple program as helpers or sole linters.
+
+coala will separated into two parts: The coala Engine and The coala CLI.
```

Comment:
Instead of one coala process, there will be two separate processes in the new architecture, is my understanding correct? Maybe this paragraph can be elaborated a bit more :)